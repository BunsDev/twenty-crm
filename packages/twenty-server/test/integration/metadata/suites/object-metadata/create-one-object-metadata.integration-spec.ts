import { ErrorCode } from 'src/engine/core-modules/graphql/utils/graphql-errors.util';
import { CreateObjectInput } from 'src/engine/metadata-modules/object-metadata/dtos/create-object.input';
import { createOneObjectMetadataFactory } from 'test/integration/metadata/suites/object-metadata/utils/create-one-object-metadata-factory.util';
import { deleteOneObjectMetadataItem } from 'test/integration/metadata/suites/object-metadata/utils/delete-one-object-metadata.util';
import { makeMetadataAPIRequest } from 'test/integration/metadata/suites/utils/make-metadata-api-request.util';
import { EachTestingContext, isDefined } from 'twenty-shared';

type CreateObjectInputPayload = Omit<
  CreateObjectInput,
  'workspaceId' | 'dataSourceId'
>;
const FailingCreationTestsUseCase: EachTestingContext<
  Partial<CreateObjectInputPayload>
>[] = [
  {
    title: 'when nameSingular has invalid characters',
    context: { nameSingular: 'μ' },
  },
  {
    title: 'when namePlural has invalid characters',
    context: { namePlural: 'μ' },
  },
  {
    title: 'when nameSingular is a reserved keyword',
    context: { nameSingular: 'user' },
  },
  {
    title: 'when namePlural is a reserved keyword',
    context: { namePlural: 'users' },
  },
  {
    title: 'when nameSingular is not camelCased',
    context: { nameSingular: 'Not_Camel_Case' },
  },
  {
    title: 'when namePlural is not camelCased',
    context: { namePlural: 'Not_Camel_Case' },
  },
  {
    title: 'when namePlural is an empty string',
    context: { namePlural: '' },
  },
  {
    title: 'when nameSingular is an empty string',
    context: { nameSingular: '' },
  },
  {
    title: 'when name exceeds maximum length',
    context: { nameSingular: 'a'.repeat(64) },
  },
];

const getMockCreateObjectInput = (
  overrides?: Partial<CreateObjectInputPayload>,
): CreateObjectInputPayload => ({
  namePlural: 'listings',
  nameSingular: 'toto',
  labelPlural: 'Listings',
  labelSingular: 'Listing',
  description: 'Listing object',
  icon: 'IconListNumbers',
  isLabelSyncedWithName: false,
  ...overrides,
});

describe('Object metadata creation should fail', () => {
  const performFailingObjectMetadataCreation = async (
    args: CreateObjectInputPayload,
  ) => {
    const graphqlOperation = createOneObjectMetadataFactory({
      input: { object: args },
      gqlFields: `
          id
          nameSingular
      `,
    });

    const response = await makeMetadataAPIRequest(graphqlOperation);
    if (isDefined(response.body.data)) {
      try {
        const createdId = response.body.data.createOneObject.id;
        await deleteOneObjectMetadataItem(createdId);
      } catch (e) {
        console.error(e);
      }
      fail('Object Metadata Item should have failed but did not');
    }
    expect(response.body.errors.length).toBe(1);
    return response.body.errors[0];
  };
  it.each(FailingCreationTestsUseCase)('$title', async ({ context }) => {
    const error = await performFailingObjectMetadataCreation(
      getMockCreateObjectInput(context),
    );

    expect(error.extensions.code).toBe(ErrorCode.BAD_USER_INPUT);
    expect(error.message).toMatchSnapshot();
  });
});
