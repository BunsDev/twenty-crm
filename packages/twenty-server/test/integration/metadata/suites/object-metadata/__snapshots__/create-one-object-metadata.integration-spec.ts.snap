// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Object metadata creation should fail when namePlural has invalid characters 1`] = `"Characters used in name "μ" are not supported"`;

exports[`Object metadata creation should fail when namePlural is a reserved keyword 1`] = `"The name "users" is not available"`;

exports[`Object metadata creation should fail when namePlural is an empty string 1`] = `"Name is too short"`;

exports[`Object metadata creation should fail when namePlural is not camelCased 1`] = `"Name should be in camelCase: Not_Camel_Case"`;

exports[`Object metadata creation should fail when nameSingular has invalid characters 1`] = `"Characters used in name "μ" are not supported"`;

exports[`Object metadata creation should fail when nameSingular is a reserved keyword 1`] = `"The name "user" is not available"`;

exports[`Object metadata creation should fail when nameSingular is an empty string 1`] = `"Name is too short"`;

exports[`Object metadata creation should fail when nameSingular is not camelCased 1`] = `"Name should be in camelCase: Not_Camel_Case"`;
